# Problem Statement
#### Analysis of laptops across various platforms to understand the factors influencing the price and specifications influencing customers interest in purchasing a laptop.
#### Through this analysis we can find out the elements that are important in fixing the prices, and the elements that makes the major contribution in the price. Also finding the main features of the laptop that are prior requirements of the customers , and which are making a clear impression on the product to the customer.
#### This analysis will help both seller and customer to obtain the best beniefits as for seller the information about the requirements and the first things that customer observe while before the buying the laptop will help making better prices and better products according to the customers requirements.For customers it would be helpful in chossing the right product acording to their price range and the best features that can be obtained.
#### For this we decieded to obtain data from online ecommerce weebsites Amazon,Flipkart,BestBuy

# Data Collection and Cleaning - Flipkart 
- The dataset was scrapped from flipkart website.
- Scraping and Cleaning done by:
    - Name: Vaibahv Saran
    - UB ID: 50615031 

# Data Collection - Flipkart

## 1. Importing Modules

In [1]:
import requests
from bs4 import BeautifulSoup
import re
import numpy as np
import pandas as pd
import time
import random

### Explanation for Modules
- `requests` module is to send a request to the URL to fetch the data.
- `BeautifulSoup` is a class using the object of which we will deal with the scraped HTML data.
- `re` is for using regex patterns to filter out data and create our dataframe in an organized format.
- `numpy` and `pandas` module if for manipulating data values and handling the data overall.
- `time` module is used to create a time delay during scraping.
- `random` module is used to generate random numbers to be used during scraping time delays.

## 2. Scraping All The Webpages Of Flipkart For Laptop Data

<h3><a href="https://www.flipkart.com/search?q=laptop&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off&page=1">Click here for the Initial Site</a></h3>

![image.jpg](images\pages.jpg)

- As highlighted in the red box of the above image we have access to **68 pages** of flipkart to scrape the data from.
- So based on that we will write the code to scrape the data

In [2]:
# Defining Request Headers to scrape the data
request_header = {
    'Content-Type': 'text/html; charset=UTF-8',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/119.0',
    'Accept-Encoding': 'gzip, deflate, br',
    'Referer': 'https://www.flipkart.com/',
    'Origin': 'https://www.flipkart.com',
    'Accept-Language': 'en-US,en;q=0.9'
}

### Understanding The Request Headers

- **Content-Type**: This tells the server what kind of data you’re sending. In this case, it’s HTML text with a specific character set (UTF-8).
- **User-Agent**: This identifies the browser and operating system making the request. It helps the server understand how to format the response. For example, we are using Firefox on a Windows 10 machine.
- **Accept-Encoding**: This tells the server which compression methods your client can handle. Here, it indicates that the client can accept responses compressed with gzip, deflate, or br (Brotli).
- **Referer**: This indicates the URL from which the request originated. It helps the server understand the context of the request. In this scenario, it shows that the request is coming from Flipkart’s website.
- **Origin**: Similar to Referer, this specifies the origin of the request, which is also Flipkart in this case. It’s used for security purposes, particularly with cross-origin requests.
- **Accept-Language**: This tells the server which languages your client prefers. Here, it indicates a preference for US English, but can also accept other forms of English.

In [3]:
# Storing the URL as a f-string
page = 1
URL = f"https://www.flipkart.com/search?q=laptop&otracker=search&otracker1=search&marketplace=FLIPKART&as-show=on&as=off&page={page}"

- The URL is stored as f-string becuase, by changing the page number in the URL, we can access the next page data, this can be utilized in conjunction with for loop to scrape all the available data.

In [4]:
# Scraping Code

total_pages = 68 # Total number of pages being scraped
i = 1 # Counter to self verify the pages being scraped successfully
raw_text = [] # List to store all the raw html code

# Loop to iterate over all the pages by changing the f-string URL
for page in range (1, total_pages+1):

    # Fetching the data from URL based on the above request headers
    response = requests.get(URL, headers=request_header)

    # Random number to be used as time delay in order to make the script behaviour more human like
    delay = random.randint(5,10)
    print("Time Delay:",delay,end=" seconds    : ")

    # While Loop: covers the edge case wherein the first attempt to fetch the data failed, 
    # by continuously requesting the data at irregular time intervals in order to mimic human behavior
    while response.status_code!=200:
        time.sleep(delay)
        response = requests.get(URL,headers=request_header)

    # Confirmation Message of Successful Scrape
    print("Page",i," status:",response)

    # Incrementing Page Counter
    i+=1  

    # Appending the raw HTML code in the list
    raw_text.append(response.text)

    # A random delay before requesting the data from next page
    time.sleep(delay)

Time Delay: 9 seconds    : Page 1  status: <Response [200]>
Time Delay: 7 seconds    : Page 2  status: <Response [200]>
Time Delay: 9 seconds    : Page 3  status: <Response [200]>
Time Delay: 9 seconds    : Page 4  status: <Response [200]>
Time Delay: 6 seconds    : Page 5  status: <Response [200]>
Time Delay: 6 seconds    : Page 6  status: <Response [200]>
Time Delay: 7 seconds    : Page 7  status: <Response [200]>
Time Delay: 9 seconds    : Page 8  status: <Response [200]>
Time Delay: 8 seconds    : Page 9  status: <Response [200]>
Time Delay: 10 seconds    : Page 10  status: <Response [200]>
Time Delay: 9 seconds    : Page 11  status: <Response [200]>
Time Delay: 7 seconds    : Page 12  status: <Response [200]>
Time Delay: 9 seconds    : Page 13  status: <Response [200]>
Time Delay: 7 seconds    : Page 14  status: <Response [200]>
Time Delay: 5 seconds    : Page 15  status: <Response [200]>
Time Delay: 7 seconds    : Page 16  status: <Response [200]>
Time Delay: 10 seconds    : Page

## 3. Saving The Raw HTML Data in CSV

- Now we will save the raw HTML code for each page in a CSV by converting the list into a dataframe.
- Saving in CSV will ensure that we don't have to scrape the entire data everytime we want to work on the data as scraping itself is a time consuming process.

In [5]:
# Converting the list to Data Frame
df = pd.DataFrame(raw_text,columns=["Raw Data"])

# Printing a sample to ensure correct data format
df.head()

Raw Data
0  <!doctype html><html lang="en"><head><link hre...
1  <!doctype html><html lang="en"><head><link hre...
2  <!doctype html><html lang="en"><head><link hre...
3  <!doctype html><html lang="en"><head><link hre...
4  <!doctype html><html lang="en"><head><link hre...

In [6]:
# Dataframe has been created successfully and can now be saved in a CSV file
df.to_csv(r"data\raw.csv")

## Data Cleaning - Flipkart

## 1. Importing Requisite Libraries

In [7]:
import numpy as np
import pandas as pd
import re
from bs4 import BeautifulSoup

### List Of The Features To Be Extracted From The Raw Data

- laptop_company = []
- processor_company = []
- processor = []
- operating_system = []
- RAM = []
- storage = []
- storage_type = [] # If not SSD Default will be HDD
- rating = []
- No_reviews = []
- screen_size = []
- price = [] # Target column

### 1.1 Filtering Features From Raw Data

In [8]:
# Loading the raw CSV
df = pd.read_csv(r"data\raw.csv")
df.head()

Unnamed: 0                                           Raw Data
0           0  <!doctype html><html lang="en"><head><link hre...
1           1  <!doctype html><html lang="en"><head><link hre...
2           2  <!doctype html><html lang="en"><head><link hre...
3           3  <!doctype html><html lang="en"><head><link hre...
4           4  <!doctype html><html lang="en"><head><link hre...

In [9]:
pages = []
for i,j in df.iterrows():
    pages.append(j["Raw Data"])

In [10]:
# Iterating over all pages and for each page filtering out the laptop brand for each product
laptop_brand = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="tUxRFH"):
        regex = re.findall("Compare(\w+)",i.text)
        if regex:
            laptop_brand.append(regex[0])
        else:
            laptop_brand.append(np.nan)

In [11]:
len(laptop_brand)

1632

In [12]:
# Filtering out laptop names
laptop_name = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):
        regex = re.findall("(^.+)\s(?:Intel|intel|AMD|M1|M2|M3|Chromebook|Snapdragon)",i.text)
        if regex:
            laptop_name.append(regex[0])
        else:
            laptop_name.append(np.nan)

In [13]:
len(laptop_name)

1632

In [14]:
processor = []
processor_company = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):        
        
        # Regex to find the processor company of the laptop
        regex1 = re.findall("Intel|intel|AMD|M1|M2|M3|Chromebook|Snapdragon",str(i.text))
        if regex1:
            processor_company.append(regex1[0])
        else:
            processor_company.append(np.nan)

        # Regex to find the exact processor in the laptop
        regex2 = re.findall("(?:Intel|intel|AMD|M1|M2|M3|Chromebook|Snapdragon)\s(.+) - ",str(i.text))
        if regex2:
            processor.append(regex2[0])
        else:
            processor.append(np.nan)

In [15]:
# Checking the number of datapoints after applying the regex
print(len(processor_company))

1632


In [16]:
print(len(processor))

1632


In [17]:
# Filtering out operating System of the laptops
operating_system = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):

        # Regex to find Operating Systems
        regex = re.findall(".+((?:Windows 10|Mac OS|DOS|Andorid|Chrome|Windows 11|Windows 11 Home))",str(i.text))

        if regex:
            operating_system.append(regex[0])
        else:
            operating_system.append(np.nan)

In [18]:
# Checking the number of datapoints
len(operating_system)

1632

In [19]:
# Filtering out RAM of the laptops
RAM = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):

        # Regex to find RAM
        regex = re.findall("(\d+)\sGB\/",str(i.text))

        if regex:
            RAM.append(regex[0])
        else:
            RAM.append(np.nan)

In [20]:
len(RAM)

1632

In [21]:
# Filtering out Storage of the laptops
storage = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):

        # Regex to find Storage Size
        regex = re.findall("\d+\sGB\/(\d+)\s(?:GB|TB)\s(?:SSD|HDD|EMMC)",str(i.text))

        if regex:
            storage.append(regex[0])
        else:
            storage.append(np.nan)

In [22]:
len(storage)

1632

In [23]:
# Filtering out Storage type of the laptops
storage_type = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="KzDlHZ"):

        # Regex to find Storage type
        regex = re.findall("\d+\sGB\/\d+\s(?:GB|TB)\s((?:SSD|HDD|EMMC))",str(i.text))

        if regex:
            storage_type.append(regex[0])
        else:
            storage_type.append(np.nan)

In [24]:
len(storage_type)

1632

In [25]:
# Filtering out Rating of the laptops
rating = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="tUxRFH"):

        r = i.find("div",class_="XQDdHH")

        if r:
            rating.append(r.text)
        else:
            rating.append(np.nan)

In [26]:
len(rating)

1632

In [27]:
# Filtering out number of reviews for each laptops
No_reviews = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="tUxRFH"):
        p =i.find("span",class_="Wphh3N")
        if p:
            regex = re.findall("\&\s(.+)\sReviews",p.text)
            if regex:
                No_reviews.append(regex[0])
            else:
                No_reviews.append(np.nan)
        else:
            No_reviews.append(np.nan)

In [28]:
len(No_reviews)

1632

In [29]:
# Filtering out Screen Size of the laptops
screen_size = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="_6NESgJ"):
        regex = re.findall("\d+?\.?\d+?",i.text)
        if regex:
            screen_size.append(regex[0])
        else:
            screen_size.append(np.nan)

In [30]:
len(screen_size)

1632

In [31]:
# Filtering out the target column: price of the laptop
price = []
for page in pages:
    soup = BeautifulSoup(page)
    for i in soup.find_all("div",class_="tUxRFH"):
        
        p =i.find("div",class_="Nx9bqj _4b5DiR")

        if p:
            price.append(p.text)
        else:
            price.append(np.nan)

In [32]:
len(price)

1632

In [33]:
# Creating a dataframe from all the extracted features present in list
feature_dict = {"Laptop_Brand":laptop_brand,
                "Laptop_Name":laptop_name,
                "Processor_Company":processor_company,
                "Processor":processor,
                "Operating_System":operating_system,
                "RAM":RAM,
                "Storage":storage,
                "Storage_Type":storage_type,
                "Screen_Size":screen_size,
                "Rating":rating,
                "Number_of_Reviews": No_reviews,
                "Price":price}

In [34]:
# Creating a dataframe from the above dictionary
laptop_df = pd.DataFrame(feature_dict)

In [35]:
# Saving the dataframe as a csv file for further analysis
laptop_df.to_csv(r"data\flipkart_laptop_data.csv",index=False)

# Data Cleaning - Flipkart

## 1. Loading the Necessary Modules

In [36]:
import numpy as np
import pandas as pd
import re
import warnings
warnings.filterwarnings("ignore")

## 2. Loading the CSV file

In [37]:
laptop_df = pd.read_csv(r"data\flipkart_laptop_data.csv")
laptop_df.head()

Laptop_Brand  Laptop_Name Processor_Company                Processor  \
0           HP    HP Victus             Intel         Core i5 12th Gen   
1          MSI  MSI Thin 15             Intel  Core i5 12th Gen 12450H   
2           HP    HP Laptop               AMD  Ryzen 3 Quad Core 5300U   
3         Acer     Acer One             Intel  Core i3 11th Gen 1115G4   
4           HP           HP               AMD  Ryzen 5 Hexa Core 5500U   

  Operating_System  RAM  Storage Storage_Type  Screen_Size  Rating  \
0       Windows 11   16      512          SSD           12     4.4   
1       Windows 11   16      512          SSD           12     4.3   
2       Windows 11    8      512          SSD           11     4.3   
3       Windows 11    8      512          SSD           11     4.2   
4       Windows 11   16      512          SSD           16     4.3   

   Number_of_Reviews    Price  
0               38.0  ₹58,990  
1               34.0  ₹57,990  
2              482.0  ₹30,999  
3              571.0  ₹26,990  
4              268.0  ₹42,990

## 3. Cleaning of Dataset

In [38]:
laptop_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1632 entries, 0 to 1631
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Laptop_Brand       1632 non-null   object 
 1   Laptop_Name        1632 non-null   object 
 2   Processor_Company  1632 non-null   object 
 3   Processor          1632 non-null   object 
 4   Operating_System   1632 non-null   object 
 5   RAM                1632 non-null   int64  
 6   Storage            1632 non-null   int64  
 7   Storage_Type       1632 non-null   object 
 8   Screen_Size        1632 non-null   int64  
 9   Rating             1503 non-null   float64
 10  Number_of_Reviews  1503 non-null   float64
 11  Price              1623 non-null   object 
dtypes: float64(2), int64(3), object(7)
memory usage: 153.1+ KB


### Observation
- There are a total of `1632` datapoints but looking at the columns `Rating` and `Number_of_Reviews`, there are some null values which needs to be dealt with later.
- Need to check for wrong values or outliers in the data.
- The target column `Price` should be integer but is stored as an object so it must be converted to right datatype as well as the missing data needs to be replaced.

In [39]:
# Inspecting Price Column
laptop_df["Price"].head(10)

0    ₹58,990
1    ₹57,990
2    ₹30,999
3    ₹26,990
4    ₹42,990
5    ₹64,990
6    ₹52,990
7    ₹35,990
8    ₹20,990
9    ₹36,990
Name: Price, dtype: object

### Observation
- Based on the above cell output, we see that Price is being treated as object column becuase of an extra symbol `₹` and `,`.
- Therefore they need to be removed as they dont contribute in the analysis.

In [40]:
# Removing extra characters from the price column
laptop_df["Price"] = laptop_df["Price"].str.replace(',','').str.replace('₹', '')
laptop_df.head()

Laptop_Brand  Laptop_Name Processor_Company                Processor  \
0           HP    HP Victus             Intel         Core i5 12th Gen   
1          MSI  MSI Thin 15             Intel  Core i5 12th Gen 12450H   
2           HP    HP Laptop               AMD  Ryzen 3 Quad Core 5300U   
3         Acer     Acer One             Intel  Core i3 11th Gen 1115G4   
4           HP           HP               AMD  Ryzen 5 Hexa Core 5500U   

  Operating_System  RAM  Storage Storage_Type  Screen_Size  Rating  \
0       Windows 11   16      512          SSD           12     4.4   
1       Windows 11   16      512          SSD           12     4.3   
2       Windows 11    8      512          SSD           11     4.3   
3       Windows 11    8      512          SSD           11     4.2   
4       Windows 11   16      512          SSD           16     4.3   

   Number_of_Reviews  Price  
0               38.0  58990  
1               34.0  57990  
2              482.0  30999  
3              571.0  26990  
4              268.0  42990

In [41]:
# Analysing the complete description summary of the dataframe
laptop_df.describe(include="all").T

count unique                     top  freq        mean  \
Laptop_Brand         1632      9                      HP   388         NaN   
Laptop_Name          1632     49                      HP   136         NaN   
Processor_Company    1632      4                   Intel  1040         NaN   
Processor            1632     38  Core i3 12th Gen 1215U   182         NaN   
Operating_System     1632      2              Windows 11  1541         NaN   
RAM                1632.0    NaN                     NaN   NaN   12.252451   
Storage            1632.0    NaN                     NaN   NaN  440.907475   
Storage_Type         1632      2                     SSD  1541         NaN   
Screen_Size        1632.0    NaN                     NaN   NaN   19.865809   
Rating             1503.0    NaN                     NaN   NaN    4.203127   
Number_of_Reviews  1503.0    NaN                     NaN   NaN  246.401863   
Price                1623     53                   54990   100         NaN   

                          std   min    25%    50%    75%    max  
Laptop_Brand              NaN   NaN    NaN    NaN    NaN    NaN  
Laptop_Name               NaN   NaN    NaN    NaN    NaN    NaN  
Processor_Company         NaN   NaN    NaN    NaN    NaN    NaN  
Processor                 NaN   NaN    NaN    NaN    NaN    NaN  
Operating_System          NaN   NaN    NaN    NaN    NaN    NaN  
RAM                  5.865025   4.0    8.0    8.0   16.0   32.0  
Storage            171.443332   1.0  512.0  512.0  512.0  512.0  
Storage_Type              NaN   NaN    NaN    NaN    NaN    NaN  
Screen_Size         18.555163  11.0   11.0   12.0   16.0   81.0  
Rating               0.238804   3.3    4.1    4.2    4.3    5.0  
Number_of_Reviews  202.948019   0.0   25.0  214.0  467.0  597.0  
Price                     NaN   NaN    NaN    NaN    NaN    NaN

### Observation Based Tasks:
- Convert the price column to int after handling missing values.
- Convert the laptop names to proper product names
- In the storage column, there is a min vlaue of 1, which is measuring the data in TB, so all the values must be converted to a common GB measurement.
- Check the screen size for values and impute the outliers accordingly.
- In rating and no of reviews replace null values with 0.

In [42]:
# Analyzing the datapoints which have null price
laptop_df[laptop_df["Price"].isnull()]

Laptop_Brand Laptop_Name Processor_Company                 Processor  \
25         Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
145        Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
217        Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
241        Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
265        Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
641        Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
1457       Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
1553       Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   
1577       Lenovo  Lenovo LOQ             Intel  Core i5 13th Gen 13450HX   

     Operating_System  RAM  Storage Storage_Type  Screen_Size  Rating  \
25         Windows 11   16      512          SSD           13     4.2   
145        Windows 11   16      512          SSD           13     4.2   
217        Windows 11   16      512          SSD           13     4.2   
241        Windows 11   16      512          SSD           13     4.2   
265        Windows 11   16      512          SSD           13     4.2   
641        Windows 11   16      512          SSD           13     4.2   
1457       Windows 11   16      512          SSD           13     4.2   
1553       Windows 11   16      512          SSD           13     4.2   
1577       Windows 11   16      512          SSD           13     4.2   

      Number_of_Reviews Price  
25                 44.0   NaN  
145                44.0   NaN  
217                44.0   NaN  
241                44.0   NaN  
265                44.0   NaN  
641                44.0   NaN  
1457               44.0   NaN  
1553               44.0   NaN  
1577               44.0   NaN

### Observation
- All the above data points are duplicated, so will be dropped.
- The process of handling will be as follows:
    - Removing all the null values in Price column excpet for the 1 datapoint.
    - Replacing the null value in Price with the mean of 50th percentile and 75th percentile of all Lenovo Laptop Prices. 

In [43]:
# Droping 8 rows where Laptop_Name is 'Lenovo LOQ' and Price is null
lenovo_loq_null_price = laptop_df[(laptop_df['Laptop_Name'] == 'Lenovo LOQ') & (laptop_df['Price'].isnull())]
laptop_df = laptop_df.drop(lenovo_loq_null_price.index[:8])

# Calculating the mean of the 50th and 75th percentiles of the Price column for Lenovo laptops
lenovo_prices = laptop_df[laptop_df['Laptop_Brand'] == 'Lenovo']['Price'].dropna().astype(float)
percentile_50 = lenovo_prices.quantile(0.50)
percentile_75 = lenovo_prices.quantile(0.75)
mean_price = (percentile_50 + percentile_75) / 2

# Replacing the null value in the remaining 'Lenovo LOQ' row with the calculated mean price
laptop_df.loc[(laptop_df['Laptop_Name'] == 'Lenovo LOQ') & (laptop_df['Price'].isnull()), 'Price'] = mean_price

# Verifying the changes
laptop_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1624 entries, 0 to 1631
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Laptop_Brand       1624 non-null   object 
 1   Laptop_Name        1624 non-null   object 
 2   Processor_Company  1624 non-null   object 
 3   Processor          1624 non-null   object 
 4   Operating_System   1624 non-null   object 
 5   RAM                1624 non-null   int64  
 6   Storage            1624 non-null   int64  
 7   Storage_Type       1624 non-null   object 
 8   Screen_Size        1624 non-null   int64  
 9   Rating             1495 non-null   float64
 10  Number_of_Reviews  1495 non-null   float64
 11  Price              1624 non-null   object 
dtypes: float64(2), int64(3), object(7)
memory usage: 164.9+ KB


In [44]:
# Converting Price column to numbers
laptop_df["Price"] = laptop_df["Price"].astype("float")
laptop_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1624 entries, 0 to 1631
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Laptop_Brand       1624 non-null   object 
 1   Laptop_Name        1624 non-null   object 
 2   Processor_Company  1624 non-null   object 
 3   Processor          1624 non-null   object 
 4   Operating_System   1624 non-null   object 
 5   RAM                1624 non-null   int64  
 6   Storage            1624 non-null   int64  
 7   Storage_Type       1624 non-null   object 
 8   Screen_Size        1624 non-null   int64  
 9   Rating             1495 non-null   float64
 10  Number_of_Reviews  1495 non-null   float64
 11  Price              1624 non-null   float64
dtypes: float64(3), int64(3), object(6)
memory usage: 164.9+ KB


In [45]:
# Removing the brand name from Laptop Name to just have product name
def remove_company(name):
    words = name.split(' ',1)
    
    if len(words)>1:
        # Removing anything which comes after , or -
        trunc = words[1].split(',')[0].split('-')[0].strip()
        return trunc if trunc else name
        
    return name

laptop_df["Laptop_Name"] = laptop_df["Laptop_Name"].apply(remove_company)

In [46]:
laptop_df.head(10)

Laptop_Brand  Laptop_Name Processor_Company                Processor  \
0           HP       Victus             Intel         Core i5 12th Gen   
1          MSI      Thin 15             Intel  Core i5 12th Gen 12450H   
2           HP       Laptop               AMD  Ryzen 3 Quad Core 5300U   
3         Acer          One             Intel  Core i3 11th Gen 1115G4   
4           HP           HP               AMD  Ryzen 5 Hexa Core 5500U   
5      Infinix      GT Book             Intel  Core i5 12th Gen 12450H   
6         Acer     Aspire 7             Intel  Core i5 12th Gen 12450H   
7         ASUS  Vivobook 15             Intel   Core i3 12th Gen 1215U   
8         Acer     Aspire 3             Intel  Celeron Dual Core N4500   
9          MSI    Modern 14               AMD  Ryzen 5 Hexa Core 7530U   

  Operating_System  RAM  Storage Storage_Type  Screen_Size  Rating  \
0       Windows 11   16      512          SSD           12     4.4   
1       Windows 11   16      512          SSD           12     4.3   
2       Windows 11    8      512          SSD           11     4.3   
3       Windows 11    8      512          SSD           11     4.2   
4       Windows 11   16      512          SSD           16     4.3   
5       Windows 11   16      512          SSD           12     4.4   
6       Windows 11   16      512          SSD           12     4.1   
7       Windows 11    8      512          SSD           12     4.2   
8       Windows 11    8      512          SSD           11     3.8   
9       Windows 11   16      512          SSD           16     4.3   

   Number_of_Reviews    Price  
0               38.0  58990.0  
1               34.0  57990.0  
2              482.0  30999.0  
3              571.0  26990.0  
4              268.0  42990.0  
5               13.0  64990.0  
6              214.0  52990.0  
7              360.0  35990.0  
8               25.0  20990.0  
9              246.0  36990.0

In [47]:
# Handling the values of Storage column
laptop_df[["Storage"]].describe()

Storage
count  1624.000000
mean    440.557266
std     171.792497
min       1.000000
25%     512.000000
50%     512.000000
75%     512.000000
max     512.000000

### Observation
- The min value of a laptop storage can be 128 GB nothing less than that, if it is less than it means the value is TB and needs to be converted into GB or its a wrong data point.
- Lets filter out all the datapoints where the storage is less than 128 GB.

In [48]:
# Filtering out the datapoints where the storage is less than 128
filtered_df = laptop_df[laptop_df['Storage'] < 128]
filtered_df

Laptop_Brand                  Laptop_Name Processor_Company  \
17           ASUS            ROG Strix Scar 16             Intel   
18             HP   Chromebook MediaTek MT8183        Chromebook   
20           Acer                 Predator Neo             Intel   
29            MSI                   Claw AI PC             Intel   
41           Acer  Acer Predator Helios Neo 16             Intel   
...           ...                          ...               ...   
1601       Lenovo        Yoga Slim 7x Qualcomm        Snapdragon   
1602           HP   Chromebook MediaTek MT8183        Chromebook   
1613       Lenovo                   Yoga AI PC             Intel   
1625       Lenovo        Yoga Slim 7x Qualcomm        Snapdragon   
1626           HP   Chromebook MediaTek MT8183        Chromebook   

                     Processor Operating_System  RAM  Storage Storage_Type  \
17    Core i9 14th Gen 14900HX       Windows 11   32        2          SSD   
18             MediaTek MT8183           Chrome    4       32         EMMC   
20    Core i7 13th Gen 13700HX       Windows 11   16        1          SSD   
29           Core Ultra 7 155H       Windows 11   16        1          SSD   
41    Core i9 13th Gen 13900HX       Windows 11   16        1          SSD   
...                        ...              ...  ...      ...          ...   
1601                   X Elite       Windows 11   32        1          SSD   
1602           MediaTek MT8183           Chrome    4       32         EMMC   
1613         Core Ultra 7 155H       Windows 11   32        1          SSD   
1625                   X Elite       Windows 11   32        1          SSD   
1626           MediaTek MT8183           Chrome    4       32         EMMC   

      Screen_Size  Rating  Number_of_Reviews     Price  
17             14     NaN                NaN  339990.0  
18             81     3.8              501.0   11990.0  
20             13     4.4               89.0  104990.0  
29             16     5.0                1.0   74990.0  
41             13     4.2                8.0  134990.0  
...           ...     ...                ...       ...  
1601           32     NaN                NaN  149990.0  
1602           81     3.8              501.0   11990.0  
1613           32     NaN                NaN  244890.0  
1625           32     NaN                NaN  149990.0  
1626           81     3.8              501.0   11990.0  

[223 rows x 12 columns]

### Observation
- EMMC Storage are exception to the traditional laptops as they aare made for extremely light weight load so whatever storage is provided need not to be changed.
- However the storage value in for SSDs/HDDs needs to be updated.

In [49]:
# Converting the Storage values of Storage in TB to GB
condition = (laptop_df['Storage_Type'].isin(['SSD', 'HDD'])) & (laptop_df['Storage'] < 128)
laptop_df.loc[condition, 'Storage'] *= 1024

In [50]:
# Looking at Data Sumamry
laptop_df.describe(include="all").T

count unique                     top  freq          mean  \
Laptop_Brand         1624      9                      HP   388           NaN   
Laptop_Name          1624     47                      HP   136           NaN   
Processor_Company    1624      4                   Intel  1032           NaN   
Processor            1624     38  Core i3 12th Gen 1215U   182           NaN   
Operating_System     1624      2              Windows 11  1533           NaN   
RAM                1624.0    NaN                     NaN   NaN      12.23399   
Storage            1624.0    NaN                     NaN   NaN    526.857143   
Storage_Type         1624      2                     SSD  1533           NaN   
Screen_Size        1624.0    NaN                     NaN   NaN     19.899631   
Rating             1495.0    NaN                     NaN   NaN      4.203144   
Number_of_Reviews  1495.0    NaN                     NaN   NaN     247.48495   
Price              1624.0    NaN                     NaN   NaN  53425.598522   

                           std      min      25%      50%      75%       max  
Laptop_Brand               NaN      NaN      NaN      NaN      NaN       NaN  
Laptop_Name                NaN      NaN      NaN      NaN      NaN       NaN  
Processor_Company          NaN      NaN      NaN      NaN      NaN       NaN  
Processor                  NaN      NaN      NaN      NaN      NaN       NaN  
Operating_System           NaN      NaN      NaN      NaN      NaN       NaN  
RAM                   5.873543      4.0      8.0      8.0     16.0      32.0  
Storage             202.351364     32.0    512.0    512.0    512.0    2048.0  
Storage_Type               NaN      NaN      NaN      NaN      NaN       NaN  
Screen_Size          18.594559     11.0     11.0     12.0     16.0      81.0  
Rating                0.239443      3.3      4.1      4.2      4.3       5.0  
Number_of_Reviews   202.948047      0.0     25.0    214.0    467.0     597.0  
Price              49743.83399  11990.0  30999.0  37999.0  54990.0  339990.0

### Observation
- Based on the above describe(Mean, precentiles and median) of the storage column, we can be sure that all the values are now accurate.

In [51]:
# Dealing with Screen Size outlier values
laptop_df[laptop_df["Screen_Size"] > 17].head()

Laptop_Brand                 Laptop_Name Processor_Company  \
12           HP                         15s               AMD   
18           HP  Chromebook MediaTek MT8183        Chromebook   
36           HP                         15s               AMD   
42           HP  Chromebook MediaTek MT8183        Chromebook   
60           HP                         15s               AMD   

                  Processor Operating_System  RAM  Storage Storage_Type  \
12  Ryzen 3 Quad Core 5300U       Windows 11    8      512          SSD   
18          MediaTek MT8183           Chrome    4       32         EMMC   
36  Ryzen 3 Quad Core 5300U       Windows 11    8      512          SSD   
42          MediaTek MT8183           Chrome    4       32         EMMC   
60  Ryzen 3 Quad Core 5300U       Windows 11    8      512          SSD   

    Screen_Size  Rating  Number_of_Reviews    Price  
12           64     4.2              405.0  32490.0  
18           81     3.8              501.0  11990.0  
36           64     4.2              405.0  32490.0  
42           81     3.8              501.0  11990.0  
60           64     4.2              405.0  32490.0

In [52]:
laptop_df[laptop_df["Screen_Size"] > 17].shape

(265, 12)

### Observation
- The Screen Size values have been mislabelled during the aggregation or were not present in the give data.
- In order to impute these outlier values, these values will be replaced by the mean value of screen size based on each laptop company

In [53]:
# Filtering out what all values are there in Screen Size
screen_size_counts = laptop_df['Screen_Size'].value_counts().sort_index()

In [54]:
screen_size_counts

Screen_Size
11    485
12    403
13    218
14      5
16    248
25     22
32     62
64    113
81     68
Name: count, dtype: int64

### Observation
- The above values confirm that some values are mis-represented, so we will be replacing it with the mean value based on each laptop.
- This will be done for screen size greater than 32, and for values between 17 and 32 will be converted to inches.

In [55]:
# Imputing outliers in Screen_Size Column

# Function to replace screen sizes greater than 32 with the median screen size for each brand
# and convert screen sizes between 17 and 32 from cm to inches
def update_screen_size(group):
    median_size = group['Screen_Size'].median()
    group.loc[group['Screen_Size'] > 32, 'Screen_Size'] = median_size
    group.loc[(group['Screen_Size'] > 17) & (group['Screen_Size'] <= 32), 'Screen_Size'] *= 0.393701
    return group

# Apply the function to each group of Laptop_Brand
laptop_df = laptop_df.groupby('Laptop_Brand').apply(update_screen_size).reset_index(drop=True)

In [56]:
# Filtering out what all values are there in Screen Size
screen_size_counts = laptop_df['Screen_Size'].value_counts().sort_index()
screen_size_counts

Screen_Size
9.842525      22
11.000000    507
12.000000    403
12.598432     62
13.000000    218
14.000000      5
16.000000    407
Name: count, dtype: int64

### Observation
- Need to replace all the screen size values less than 11 to 11 inches
- Make the the screen size to definitive 12.5 inches

In [57]:
# Replace specific Screen_Size values
laptop_df['Screen_Size'] = laptop_df['Screen_Size'].replace({9.842525: 11.00, 12.598432: 12.5})

# Filtering out what all values are there in Screen Size
screen_size_counts = laptop_df['Screen_Size'].value_counts().sort_index()
screen_size_counts

Screen_Size
11.0    529
12.0    403
12.5     62
13.0    218
14.0      5
16.0    407
Name: count, dtype: int64

### Observation
- All the screen sizes are now valid.

In [58]:
laptop_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1624 entries, 0 to 1623
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Laptop_Brand       1624 non-null   object 
 1   Laptop_Name        1624 non-null   object 
 2   Processor_Company  1624 non-null   object 
 3   Processor          1624 non-null   object 
 4   Operating_System   1624 non-null   object 
 5   RAM                1624 non-null   int64  
 6   Storage            1624 non-null   int64  
 7   Storage_Type       1624 non-null   object 
 8   Screen_Size        1624 non-null   float64
 9   Rating             1495 non-null   float64
 10  Number_of_Reviews  1495 non-null   float64
 11  Price              1624 non-null   float64
dtypes: float64(4), int64(2), object(6)
memory usage: 152.4+ KB


In [59]:
laptop_df.describe(include="all").T

count unique                     top  freq          mean  \
Laptop_Brand         1624      9                      HP   388           NaN   
Laptop_Name          1624     47                Aspire 7   136           NaN   
Processor_Company    1624      4                   Intel  1032           NaN   
Processor            1624     38  Core i3 12th Gen 1215U   182           NaN   
Operating_System     1624      2              Windows 11  1533           NaN   
RAM                1624.0    NaN                     NaN   NaN      12.23399   
Storage            1624.0    NaN                     NaN   NaN    526.857143   
Storage_Type         1624      2                     SSD  1533           NaN   
Screen_Size        1624.0    NaN                     NaN   NaN     12.836207   
Rating             1495.0    NaN                     NaN   NaN      4.203144   
Number_of_Reviews  1495.0    NaN                     NaN   NaN     247.48495   
Price              1624.0    NaN                     NaN   NaN  53425.598522   

                           std      min      25%      50%      75%       max  
Laptop_Brand               NaN      NaN      NaN      NaN      NaN       NaN  
Laptop_Name                NaN      NaN      NaN      NaN      NaN       NaN  
Processor_Company          NaN      NaN      NaN      NaN      NaN       NaN  
Processor                  NaN      NaN      NaN      NaN      NaN       NaN  
Operating_System           NaN      NaN      NaN      NaN      NaN       NaN  
RAM                   5.873543      4.0      8.0      8.0     16.0      32.0  
Storage             202.351364     32.0    512.0    512.0    512.0    2048.0  
Storage_Type               NaN      NaN      NaN      NaN      NaN       NaN  
Screen_Size            1.94802     11.0     11.0     12.0     16.0      16.0  
Rating                0.239443      3.3      4.1      4.2      4.3       5.0  
Number_of_Reviews   202.948047      0.0     25.0    214.0    467.0     597.0  
Price              49743.83399  11990.0  30999.0  37999.0  54990.0  339990.0

### Observation
- Replacing all the missing values in `Rating` and `Number_of_Reviews` with zero, as they will be treated as the laptop not being sold or not that appealing to customers.

In [60]:
# Replacing Nan values in the Rating and Number of Reviews column
laptop_df['Rating'] = laptop_df['Rating'].fillna(0)
laptop_df['Number_of_Reviews'] = laptop_df['Number_of_Reviews'].fillna(0)

In [61]:
# Taking a final look at info and description of data
laptop_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1624 entries, 0 to 1623
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Laptop_Brand       1624 non-null   object 
 1   Laptop_Name        1624 non-null   object 
 2   Processor_Company  1624 non-null   object 
 3   Processor          1624 non-null   object 
 4   Operating_System   1624 non-null   object 
 5   RAM                1624 non-null   int64  
 6   Storage            1624 non-null   int64  
 7   Storage_Type       1624 non-null   object 
 8   Screen_Size        1624 non-null   float64
 9   Rating             1624 non-null   float64
 10  Number_of_Reviews  1624 non-null   float64
 11  Price              1624 non-null   float64
dtypes: float64(4), int64(2), object(6)
memory usage: 152.4+ KB


In [62]:
laptop_df.describe(include="all").T

count unique                     top  freq          mean  \
Laptop_Brand         1624      9                      HP   388           NaN   
Laptop_Name          1624     47                Aspire 7   136           NaN   
Processor_Company    1624      4                   Intel  1032           NaN   
Processor            1624     38  Core i3 12th Gen 1215U   182           NaN   
Operating_System     1624      2              Windows 11  1533           NaN   
RAM                1624.0    NaN                     NaN   NaN      12.23399   
Storage            1624.0    NaN                     NaN   NaN    526.857143   
Storage_Type         1624      2                     SSD  1533           NaN   
Screen_Size        1624.0    NaN                     NaN   NaN     12.836207   
Rating             1624.0    NaN                     NaN   NaN      3.869273   
Number_of_Reviews  1624.0    NaN                     NaN   NaN    227.826355   
Price              1624.0    NaN                     NaN   NaN  53425.598522   

                           std      min      25%      50%      75%       max  
Laptop_Brand               NaN      NaN      NaN      NaN      NaN       NaN  
Laptop_Name                NaN      NaN      NaN      NaN      NaN       NaN  
Processor_Company          NaN      NaN      NaN      NaN      NaN       NaN  
Processor                  NaN      NaN      NaN      NaN      NaN       NaN  
Operating_System           NaN      NaN      NaN      NaN      NaN       NaN  
RAM                   5.873543      4.0      8.0      8.0     16.0      32.0  
Storage             202.351364     32.0    512.0    512.0    512.0    2048.0  
Storage_Type               NaN      NaN      NaN      NaN      NaN       NaN  
Screen_Size            1.94802     11.0     11.0     12.0     16.0      16.0  
Rating                1.159917      0.0    4.075      4.2      4.3       5.0  
Number_of_Reviews   205.902161      0.0     12.0    214.0    424.0     597.0  
Price              49743.83399  11990.0  30999.0  37999.0  54990.0  339990.0

In [63]:
# Since the data is now cleaned, it can be exported as a clean CSV for further analysis
laptop_df.to_csv(r"data\flipkart_laptop_cleaned.csv",index=False)

# Data Collection and Cleaning - Best Buy 
- The dataset was scrapped from Best Buy website.
- Scraping and Cleaning done by:
    - Name: Yeswanth Chitturi
    - UB ID: 50591666

## 1. Importing Modules

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

## 2. Scraping All The Webpages Of Best Buy For Laptop Data

<h3><a href="https://www.bestbuy.com/site/searchpage.jsp?st=laptop&_dyncharset=UTF-8&_dynSessConf=&id=pcat17071&type=page&sc=Global&cp=1&nrp=&sp=&qp=&list=n&af=true&iht=y&usc=All+Categories&ks=960&keys=keys">Click here for the Initial Site</a></h3>

## Scraping Code , Converting Html to data frame and saving as CSV.

In [2]:
# Function to get the HTML content of a page
def get_page_content(url):
    headers = {'User-Agent': 'Mozilla/5.0'}
    response = requests.get(url, headers=headers)
    return BeautifulSoup(response.content, 'html.parser')

# Function to scrape laptop data from one page
def scrape_laptop_data_from_page(soup, data):
    base_url = 'https://www.bestbuy.com'

    # Find all laptops on the page
    laptops = soup.find_all('li', class_='sku-item')

    for laptop in laptops:
        try:
            # Extract laptop name
            name_tag = laptop.find('h4', class_='sku-title')
            name = name_tag.text.strip() if name_tag else 'No name available'

            # Extract SKU value
            skuvalue_tag = laptop.find('span', class_='sku-value')
            skuvalue = skuvalue_tag.text.strip() if skuvalue_tag else 'No SKU value available'

            # Extract rating (visually hidden text)
            visually_hidden_tag = laptop.find('p', class_='visually-hidden')
            rating = visually_hidden_tag.text.strip() if visually_hidden_tag else 'No rating available'

            # Extract laptop price
            price_tag = laptop.find('div', class_='priceView-customer-price')
            price = price_tag.span.text.strip() if price_tag else 'No price available'

            # Append data to the list
            data.append([name, skuvalue, rating, price])
            #print(f'Scraped: {name}')

        except Exception as e:
           print(f'Error scraping laptop: {e}')

# Function to find the link to the next page
def get_next_page(soup):
    next_page_tag = soup.find('a', class_='sku-list-page-next')
    if next_page_tag and 'href' in next_page_tag.attrs:
        return next_page_tag['href']
    return None

# Main function to scrape all pages
def scrape_all_pages():
    base_url = 'https://www.bestbuy.com'
    search_url = f'{base_url}/site/searchpage.jsp?st=laptops'

    # List to hold all laptop data
    data = []

    current_page_url = search_url

    while current_page_url:
        #print(f'Scraping page: {current_page_url}')
        soup = get_page_content(current_page_url)
        scrape_laptop_data_from_page(soup, data)

        # Check if there's a next page
        next_page = get_next_page(soup)
        if next_page:
            current_page_url = f'{base_url}{next_page}'
        else:
            current_page_url = None

    # Create a DataFrame from the scraped data
    df = pd.DataFrame(data, columns=['total_info', 'Model No', 'Rating', 'Price'])

    # Save DataFrame to a CSV file
    df.to_csv('laptops_data.csv', index=False)
    print('Data has been saved to laptops_data.csv')
    return df

if __name__ == '__main__':
    df=scrape_all_pages()
    #print(df)
print(df.info())

Data has been saved to laptops_data.csv
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   total_info  1300 non-null   object
 1   Model No    1300 non-null   object
 2   Rating      1300 non-null   object
 3   Price       1300 non-null   object
dtypes: object(4)
memory usage: 40.8+ KB
None


## Data Collection Similar to Flipkart data
## Converting Raw Data Into Columns

In [3]:
#Extracting Columns
import pandas as pd
#Brand
df['Brand']='-'
df['Brand'] = df['total_info'].str.split(' ').str[0]
#Colur
df['Colour']='-'
df['Colour'] = df['total_info'].str.split(' ').str[-1]
#Saving original
df1=df
#Remove Special Characters
df1['total_info'] = df1['total_info'].str.replace('”', '"', regex=False)
#Ram
ram_pattern = r'(\s*\d+\s*G[Bb]\s*(Memory|RAM|Ram)\s*|(\d{1,2}GB))'
df1['ram'] = df1['total_info'].str.extract(ram_pattern, expand=False)[0]
df1['ram'] = df1['ram'].str.extract(r'(\d+)').astype(float)
df1['ram'] = df1['ram'].apply(lambda x: f"{int(x)} GB" if x < 33 else pd.NA)
#Storage
storage_pattern = r'(\d+\s*(TB|GB|G)\s*(SSD|HDD|Solid State Drive|Flash Storage|Hard Drive|eMMC|UFS|SDD|PCIe|NVMe|Storage)|\d+\s*TB\s*-\s*SSD|\d{3,4}SSD|\d{3,4}GB\s*-\s*SSD)'
df1['storage'] = df1['total_info'].str.extract(storage_pattern, expand=False)[0]
#Processor
processor_pattern = r'(Intel\s+\w+\s*\w*|\bM[12]\s+(?:Pro|Max|chip)\s*Built\s*for\s*Apple\s*-\w+|M[12]\s+(?:Pro|Max|chip)\b|M3\s+chip\s*Built\s*for\s*Apple\s*Intelligence|M3\s+\w+\s+chip\s*Built\s*for\s*Apple|M3\s+chip|AMD\s+Ryzen\s+\d+(?:-\d+|\s+\d+)\w*|AMD\s+Ryzen\s+\d+\b|AMD\s+Ryzen\s+AI\s+\d+-\d+\s+\w*|Apple\s+M1\s+\w+\s+chip\b|Core\s+\w+\s*-\d+|\bRyzen\s+\d+\s+\d+\w*|MediaTek\s+\w+\s+\d+|Snapdragon\s+\w+\s*\w*|Pentium\s+\w+\s+\d+)'
extracted_processors = df1['total_info'].str.extract(processor_pattern)
df1['processor'] = extracted_processors[0].fillna(value=pd.NA)
#Display size
display_pattern=r'(\d{2}"|\d{2}.\d"|\d{2}-inch|\d{2}-Inch|\d{2}.\d-inch|\d{2}.\d-Inch)'
extracted_displays = df1['total_info'].str.extract(display_pattern)
df1['display'] = extracted_displays[0].fillna(value=pd.NA)
#Laptop Name
model_pattern = r'(Geek Squad Certified Refurbished MacBook Air|Envy|XPS|OmniBook|ProBook|Flex|LOQ|Katana|Blade|Aero|Vector|Summit|Raider|Pulse|Cyborg|Elitebook|Precision|Galaxy book|Geek Squad Certified Refurbished Macbook®|Geek Squad Certified Refurbished MacBook Pro|MacBook Pro|GSRF MacBook Pro|MacBook Air|Refurbished MacBook®|Chromebook|ProArt P16|ProArt Studiobook|ROG Strix G16|ROG Strix SCAR|ROG Zephyrus G14|TUF A15|TUF Gaming A17|Vivobook|Vivobook Pro 15|Zenbook|Zenbook Pro Duo|Zenbook DUO Dual|Nitro 16|Predator Helios 18|Predator Triton|Swift X|TravelMate|m18 R2|m16|m18 165|Latitude 7000|Refurbished|Inspiron|Precision 3540|XPS 13|Latitude 3000|Mobile Processor|Alienware x14|Gaming Laptop IPS|EliteBook 840 G8|Elite x360 830 G11|Envy 2-in-1|Pavilion|Victus|ZBook|OMEN|gram|gram SuperSlim|ThinkPad X1|Yoga|Yoga Book|Yoga Pro|Legion|Ideapad|ThinkPad T14s|GSRF Surface|Surface Laptop - Copilot\+ PC|Surface Laptop Studio 2|Bravo 15|Commercial 14 H A13MG|Creator 17|Crosshair|Raider GE78 HX|Stealth|Summit E16 AI Studio|Vector 16 HX A14V|Blade 16|Galaxy Book2 Pro|Galaxy Book3 360|Galaxy Book4 Ultra|Geek Squad Certi Refurbis Galaxy Book Flex2 Alpha|Galaxy Book4 Edge|NEOZ3 Laptop|Galaxy Book4 Pro 360|Blade 18|ROG Zephyrus|Summit E16 AI Evo|Raider GE68 HX|Prestige|Modern 15|ThinkPad|ThinkPad T14|ThinkBook 14 G7|Spectre|AORUS|Latitude 5000|Latitude|Predator Helios|Aspire 5|ProArt PX13 13|Blade 15|ProBook 445|EliteBook 640|Raider 18|CreatorPro|Creator 16|ThinkBook|Swift 14|ROG Flow)'
df1['model'] = df1['total_info'].str.extract(model_pattern)
#Graphics
graphics_pattern = r'(NVIDIA GeForce RTX \d+|Ryzen \d+U|Intel Iris Xe Graphics|NVIDIA Quadro P1000|Intel UHD Graphics 620|NVIDIA Quadro T1000|AMD Radeon Graphics|NVIDIA GeForce RTX4070|NVIDIA Quadro P3200)'
df1['graphics'] = df1['total_info'].str.extract(graphics_pattern)
#Storage Type
storagetype_pattern = r'(FlashStorage|HardDrive|SSD|HDD|SDD|eMMC|Flash Storage|Hard Drive|NVMe|PCIe|UFS|Solid State Drive)'
df1['storage_type'] = df1['storage'].str.extract(storagetype_pattern)
#No of reviews
no_reviews_pattern = r'(\d{1,4})\s*reviews?|reviewfalse'
df1['no_reviews'] = df1['Rating'].str.extract(no_reviews_pattern)
#Rating out of 5
Rating_5_pattern = r'Rating\s*(\d(?:\.\d)?)'
df1['Rating_5'] = df1['Rating'].str.extract(Rating_5_pattern)
#Processor Company
processor_company_pattern = r'(Intel|AMD\s*Ryzen|Ryzen|Snapdragon|Core|MediaTek|M1|M2|M3)'
df1['processor_company'] = df1['processor'].str.extract(processor_company_pattern)
#Operating System
df1['os']='Windows11'
#Dropping group columns
df1 = df1.drop('total_info', axis=1)
df1 = df1.drop('Rating', axis=1)
#Saving original
df_BB=df1

## Filling Null Values

In [4]:
df_BB['graphics'] = df_BB['graphics'].fillna('No Graphics')

df_BB['no_reviews'] = df_BB['no_reviews'].fillna('0')

df_BB['Rating_5'] = df_BB['Rating_5'].fillna('0 Reviews')

df_BB['storage_type'] = df_BB['storage_type'].fillna('SSD')

df_BB['processor'] = df_BB['processor'].fillna('No Info')

df_BB['processor_company'] = df_BB['processor_company'].fillna('No Info')

df_BB['model'] = df_BB['model'].fillna('No Model')

most_frequent_display = df_BB['display'].mode()[0]
df_BB['display'] = df_BB['display'].fillna(most_frequent_display)

most_frequent_ram = df_BB['ram'].mode()[0]
df_BB['ram'] = df_BB['ram'].fillna(most_frequent_ram)

most_frequent_storage = df_BB['storage'].mode()[0]
df_BB['storage'] = df_BB['storage'].fillna(most_frequent_storage)

## Removing extra names and characters

In [5]:
#More cleaning

df_BB.loc[df_BB['Brand'] == 'Apple', 'os'] = 'Mac OS'

df_BB['processor_company'] = df_BB['processor_company'].replace({'M1': 'Apple', 'M2': 'Apple', 'M3': 'Apple','Core':'Intel','Ryzen': 'AMD Ryzen'})

df_BB['processor'] = df_BB['processor'].str.replace('Intel', '')
df_BB['processor'] = df_BB['processor'].str.replace('AMD Ryzen', '')
df_BB['processor'] = df_BB['processor'].str.replace('Snapdragon', '')
df_BB['processor'] = df_BB['processor'].str.replace('MediaTek', '')
df_BB['processor'] = df_BB['processor'].str.replace('Ryzen', '')

df_BB['Brand'] = df_BB['Brand'].str.replace('New!', '')

df_BB['display'] = df_BB['display'].str.replace('"', '')

df_BB['storage'] = df_BB['storage'].str.replace('GB', '')
df_BB['storage'] = df_BB['storage'].str.replace('SSD', '')
df_BB['storage'] = df_BB['storage'].str.replace('Solid State Drive', '')
df_BB['storage'] = df_BB['storage'].str.replace('SDD', '')
df_BB['storage'] = df_BB['storage'].str.replace('HDD', '')
df_BB['storage'] = df_BB['storage'].str.replace('TB PCIe', 'TB')
df_BB['storage'] = df_BB['storage'].str.replace('-', '')
df_BB['storage'] = df_BB['storage'].str.replace(' ', '')
df_BB['storage'] = df_BB['storage'].str.replace('HardDrive', '')
df_BB['storage'] = df_BB['storage'].str.replace('NVMe', '')
df_BB['storage'] = df_BB['storage'].str.replace('FlashStorage', '')
df_BB['storage'] = df_BB['storage'].str.replace('TB NVMe', 'TB')
df_BB['storage'] = df_BB['storage'].str.replace('Storage', '')
def convert_tb_to_gb(storage_value):
    if 'TB' in storage_value:
        
        tb_value = float(storage_value.replace('TB', '').strip())
        return str(int(tb_value * 1024))
    return storage_value
df_BB['storage'] = df_BB['storage'].apply(convert_tb_to_gb)

df_BB['storage_type'] = df_BB['storage_type'].str.replace('Solid State Drive', 'SSD')

df_BB['ram'] = df_BB['ram'].str.replace('GB', '')
df_BB['ram'] = df_BB['ram'].str.replace(' ', '')

#Merging Model columns
df_BB['processor_model'] = df_BB['processor'] + ' - ' + df_BB['Model No']
df_BB = df_BB.drop(['processor', 'Model No'], axis=1)

In [6]:
df_BB.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1300 entries, 0 to 1299
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Price              1300 non-null   object
 1   Brand              1300 non-null   object
 2   Colour             1300 non-null   object
 3   ram                1300 non-null   object
 4   storage            1300 non-null   object
 5   display            1300 non-null   object
 6   model              1300 non-null   object
 7   graphics           1300 non-null   object
 8   storage_type       1300 non-null   object
 9   no_reviews         1300 non-null   object
 10  Rating_5           1300 non-null   object
 11  processor_company  1300 non-null   object
 12  os                 1300 non-null   object
 13  processor_model    1300 non-null   object
dtypes: object(14)
memory usage: 142.3+ KB


## Renaming as required

In [7]:
#Renameing columns for better understading
df_BB = df_BB.rename(columns={'model': 'Laptop_name'})
df_BB = df_BB.rename(columns={'Price': 'Laptop_Price'})
df_BB = df_BB.rename(columns={'Brand': 'Laptop_Brand'})
df_BB = df_BB.rename(columns={'Price': 'Laptop_Price'})
df_BB = df_BB.rename(columns={'Colour': 'Laptop_Colour'})
df_BB = df_BB.rename(columns={'ram': 'Laptop_Memory_GB'})
df_BB = df_BB.rename(columns={'storage': 'Laptop_Storage_GB'})
df_BB = df_BB.rename(columns={'display': 'Laptop_Display_Size_in'})
df_BB = df_BB.rename(columns={'graphics': 'Laptop_Graphics'})
df_BB = df_BB.rename(columns={'no_reviews': 'No_Of_Reviews'})
df_BB = df_BB.rename(columns={'processor_company': 'Processor_Company_Name'})
df_BB = df_BB.rename(columns={'storage_type': 'Storage_Type'})
df_BB = df_BB.rename(columns={'os': 'Operating_System'})
df_BB = df_BB.rename(columns={'processor_model': 'Processor_Model'})

### Explanation for Modules
- `requests` module is to send a request to the URL to fetch the data.
- `BeautifulSoup` is a class using the object of which we will deal with the scraped HTML data.
- `re` is for using regex patterns to filter out data and create our dataframe in an organized format.
- `numpy` and `pandas` module if for manipulating data values and handling the data overall.

# Data Collection and Cleaning - Amazon 
- The dataset was scrapped from Amazon website.
- Scraping and Cleaning done by:
    - Name: Shaurya Mathur
    - UB ID: 

# References

1. <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status">HTML Error Codes</a>: Used for refrencing response codes and their meaning.
2. <a href="https://docs.python-requests.org/en/latest/index.html">Request Module of Python</a>: Used for creating custom headers during scraping